router/factory.go

```diff
@@ -31,14 +35,87 @@ type Factory struct {
 	ProcErrorDB      jobsdb.JobsDB
 	TransientSources transientsource.Service
 	RsourcesService  rsources.JobService
+	Logger           logger.Logger
+	Stats            stats.Stats
+
+	// Throttling
+	throttlerFactoryOnce sync.Once
+	throttlerFactory     *throttler.Factory
 }
 
 func (f *Factory) New(destination *backendconfig.DestinationT, identifier string) *HandleT {
+	f.throttlerFactoryOnce.Do(f.initThrottlerFactory)
+
 	r := &HandleT{
-		Reporting:    f.Reporting,
-		MultitenantI: f.Multitenant,
+		Reporting:        f.Reporting,
+		MultitenantI:     f.Multitenant,
+		throttlerFactory: f.throttlerFactory,
 	}
 	destConfig := getRouterConfig(destination, identifier)
 	r.Setup(f.BackendConfig, f.RouterDB, f.ProcErrorDB, destConfig, f.TransientSources, f.RsourcesService)
 	return r
 }
+
+func (f *Factory) initThrottlerFactory() {
+	var redisClient *redis.Client
+	if config.IsSet("Router.throttler.redis.addr") {
+		redisClient = redis.NewClient(&redis.Options{
+			Addr:     config.GetString("Router.throttler.redis.addr", "localhost:6379"),
+			Username: config.GetString("Router.throttler.redis.username", ""),
+			Password: config.GetString("Router.throttler.redis.password", ""),
+		})
+	}
+
+	var throttlingAlgorithm string
+	config.RegisterStringConfigVariable(
+		throttlingAlgoTypeGoRate, &throttlingAlgorithm, false, "Router.throttler.algorithm",
+	)
+
+	var (
+		err  error
+		l    *throttling.Limiter
+		opts = []throttling.Option{
+			throttling.WithStatsCollector(f.Stats),
+		}
+	)
+	switch throttlingAlgorithm {
```

Comment:
Wouldn't it be more straightforward if we slightly modified the `throttling` package's `Options` so that all necessary information was passed using a single option? e.g.

```
throttling.WithGoRate()
throttling.WithGCRA()
throttling.WithGCRABurst(burst int64)
throttling.WithRedisGCRA(redisClient *redis.Client)
throttling.WithRedisGCRABurst(redisClient *redis.Client, burst int64)
throttling.WithRedisSortedSet(redisClient)
```

Reply:
Hopefully it should be more clear now.